src/sentry/tasks/weekly_reports.py

```diff
@@ -0,0 +1,803 @@
+
+    organizations = Organization.objects.filter(status=OrganizationStatus.ACTIVE)
+    for i, organization in enumerate(
+        RangeQuerySetWrapper(organizations, step=10000, result_value_getter=lambda item: item.id)
+    ):
+        if features.has("organizations:weekly-email-refresh", organization):
+            # Create a celery task per organization
+            prepare_organization_report.delay(timestamp, duration, organization.id, dry_run=dry_run)
+
+
+# This task is launched per-organization.
+@instrumented_task(
+    name="sentry.tasks.weekly_reports.prepare_organization_report",
+    queue="reports.prepare",
+    max_retries=5,
+    acks_late=True,
+)
+def prepare_organization_report(
+    timestamp, duration, organization_id, dry_run=False, target_user=None
+):
+    organization = Organization.objects.get(id=organization_id)
+    ctx = OrganizationReportContext(timestamp, duration, organization)
+
+    # Run organization passes
+    with sentry_sdk.start_span(op="weekly_reports.user_project_ownership"):
+        user_project_ownership(ctx)
+    with sentry_sdk.start_span(op="weekly_reports.project_event_counts_for_organization"):
+        project_event_counts_for_organization(ctx)
+    with sentry_sdk.start_span(op="weekly_reports.organization_project_issue_summaries"):
+        organization_project_issue_summaries(ctx)
+
+    with sentry_sdk.start_span(op="weekly_reports.project_passes"):
+        # Run project passes
+        for project in organization.project_set.all():
+            project_key_errors(ctx, project)
+            project_key_transactions(ctx, project)
+            project_key_performance_issues(ctx, project)
+
+    with sentry_sdk.start_span(op="weekly_reports.fetch_key_error_groups"):
+        fetch_key_error_groups(ctx)
+    with sentry_sdk.start_span(op="weekly_reports.fetch_key_performance_issue_groups"):
+        fetch_key_performance_issue_groups(ctx)
+
+    # Finally, deliver the reports
+    with sentry_sdk.start_span(op="weekly_reports.deliver_reports"):
+        deliver_reports(ctx, dry_run=dry_run, target_user=target_user)
+
+
+# Organization Passes
+def user_project_ownership(ctx):
+    for (project_id, user_id) in OrganizationMember.objects.filter(
+        organization_id=ctx.organization.id, teams__projectteam__project__isnull=False
+    ).values_list("teams__projectteam__project_id", "user_id"):
+        ctx.project_ownership.setdefault(user_id, set()).add(project_id)
+
+
+def project_event_counts_for_organization(ctx):
+    def zerofill_data(data):
+        return zerofill(data, ctx.start, ctx.end, ONE_DAY, fill_default=0)
+
+    query = Query(
+        match=Entity("outcomes"),
+        select=[
+            Column("outcome"),
+            Column("category"),
+            Function("sum", [Column("quantity")], "total"),
+        ],
+        where=[
+            Condition(Column("timestamp"), Op.GTE, ctx.start),
+            Condition(Column("timestamp"), Op.LT, ctx.end + timedelta(days=1)),
```

Comment:
I think a timedelta of 1 day more to cause the data in the report vary depending at what time the task gets executed.

If a report is generated at 1AM, it will have less data than if generated at 5am.

Am I on the right track?

Reply:
That's actually not the case. ctx.end is `floor_to_utc_day(timezone.now())` and this is the same for all orgs.

Reply:
The reason it's 1 day more is because we have a granularity of 1 day, and snuba doesn't support `Op.LTE` on timestamp.